### ISSUE TITLE
NotRequired causes type error in TypedDict assignment

### ISSUE THREAD

**Issue by hyrious** [microsoft/pyright#11014]
**Describe the bug**
I'm not sure if it's a bug or my mis-understanding the typing annotation. Using `total=False` in typed dict inheritance, or `NotRequired`, causes the same shape cannot be assigned to a loose type.

**Code or Screenshots**

```python
from typing import TypedDict

class Inputs(TypedDict):
	foo: str

class Outputs(TypedDict, total=False):
	foo: str  # Can also use NotRequired[str]

def main(inputs: Inputs) -> Outputs:
	return inputs
#          ^^^^^^ Type "Inputs" is not assignable to return type "Outputs"
#                   "foo" is not required in "Outputs"  Pylance[reportReturnType]
#   return {'foo': inputs['foo']}   # This works fine.
```

**VS Code extension or command-line**

I'm using Pylance in VS Code. However this behavior can also be reproduced in [the pyright playground](https://pyright-play.net/?code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAqiApgCYAiSAxjAFC1UA2AhgM6tQCSKCArjKwAURBGUo0AlAC5aASGBgwUqKxgh6TNhwDy-PgOEkK1GABpYYGM0YBeAGLXWxaXIVKVa%2BqWLAoEZqiCqPqsytwhElAAtAB8ULowITKyIMQwvCAomDz8rLRAA).

**Comment by erictraut** [comment]
Pyright's behavior here is correct and in compliance with the Python typing spec. A value of type `Inputs` is not assignable to a value of type `Outputs`. Consider the following:

```python
i: Inputs = {"foo": ""}
o = main(i)
del o["foo"] # This is allowed because "foo" is not required
print(i["foo"]) # Crash!
```

For more details about assignability rules for TypedDict's, refer to [this chapter](https://typing.python.org/en/latest/spec/typeddict.html#assignability) of the typing spec.

**Comment by hyrious** [comment]
Thank you for the example! Interestingly the same thing can be demonstrated in TypeScript and it doesn't see it as an error:

```ts
interface Inputs { foo: () => void }
interface Outputs { foo?: () => void }

function main(inputs: Inputs): Outputs {
    return inputs
}

let i: Inputs = { foo: () => console.log('ok') }
let o = main(i)
o['foo'] = undefined
i['foo']() // Crash!
```
